functional_programming_tests/test/main.cpp

```diff
@@ -0,0 +1,250 @@
+    // What is the delta in position
+    int const del_x = goal.x - start.x;
+    int const del_y = goal.y - start.y;
+
+    // What direction to move in for each dimension
+    int const del_x_sign = std::copysign(1.0, del_x);
+    int const del_y_sign = std::copysign(1.0, del_y);
+
+    // Push start onto the path
+    Path path;
+    path.push_back(start);
+
+    auto is_occupied = [&costmap](auto const x, auto const y) -> bool {
+      return costmap.at(Position{x, y}) == 255;
+    };
+
+    // Fails if there is any obstacle in the way
+    // Move horizontally
+    for (size_t i = 0; i < (std::abs(del_x)); ++i) {
+      if (is_occupied(path.back().x + del_x_sign, path.back().y)) {
+        return std::nullopt;
+      }
+      path.push_back({path.back().x + del_x_sign, path.back().y});
+    }
+    // Move vertically
+    for (size_t i = 0; i < (std::abs(del_y)); i++) {
+      if (is_occupied(path.back().x, path.back().y + del_y_sign)) {
+        return std::nullopt;
+      }
+      path.push_back({path.back().x, path.back().y + del_y_sign});
+    }
+
+    return path;
+  }  
+
+};
+
+namespace PathGeneratorUtilities {
+
+  std::optional<Map<unsigned char>> parseSetMapRequest(const std::shared_ptr<example_srvs::srv::SetMap::Request> request ) {
+    const auto costmap = request->map;
+    Map<unsigned char> map;
+    // Get the costmap from a ros topic
+    // Check that map layout makes sense
+    if ((costmap.layout.dim[0].size * costmap.layout.dim[1].size) !=
+        costmap.layout.dim[0].stride) {
+      return std::nullopt;
+    }
+    if (costmap.layout.dim[0].stride != costmap.data.size()) {
+      return std::nullopt;
+    }
+    auto const begin = std::begin(costmap.data);
+    // Populate the map
+    for (size_t row = 0; row < costmap.layout.dim[0].size; row++) {
+      map.get_data().push_back(
+          {begin + row * costmap.layout.dim[1].size,
+           begin + (row + 1) * costmap.layout.dim[1].size});
+    }
+    return map;    
+  }
+
+  std_msgs::msg::UInt8MultiArray createUInt8MultiArrayMessageFromPath(const Path& path) {
+    auto message = std_msgs::msg::UInt8MultiArray();
+
+    message.layout.dim.resize(3, std_msgs::msg::MultiArrayDimension());
+
+    message.layout.dim[0].label = "rows";
+    message.layout.dim[0].size = path.size();
+    message.layout.dim[0].stride = path.size() * 2;
+
+    message.layout.dim[1].label = "columns";
+    message.layout.dim[1].size = 2;
+    message.layout.dim[1].stride = 1;
+
+    message.layout.dim[2].label = "channel";
+    message.layout.dim[2].size = 1;
+    message.layout.dim[2].stride = 1;
+
+    // Start pushing back the path only if there is one
+    if (path.size() > 0) {
+      for (auto const& position : path) {
+        message.data.push_back(position.x);
+        message.data.push_back(position.y);
+      }
+    }
+    return message;
+  }
+
+};
+
+Map<unsigned char> set_costmap_callback(const std::shared_ptr<example_srvs::srv::SetMap::Request> request,
```

Comment:
I hate this. Something that has two outputs (an output param and a return) doesn't seem right.
This function feels like it's only here to serve the lambda later.

Reply:
Yeah, I got rid of this function. Doesn't make sense.